### PR TITLE
[Core] Mark cms::Exception and edm::Exception functions that throw exceptions as noreturn

### PR DESCRIPTION
**Pull request by iarspider**
#### PR description:

This should hopefully solve LLVM analyzer warnings like this: [link](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-static-analysis/CMSSW_14_2_X_2024-10-13-0000/el8_amd64_gcc12/llvm-analysis/report-8c63da.html#EndPath)

#### PR validation:

Bot tests

### CODE REVIEW

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
please test

File: FWCore/Utilities/interface/ExceptionCollector.h [hunk 1]
```diff
@@ -35,7 +35,7 @@ namespace edm {
     ExceptionCollector(std::string const& initialMessage);
     ~ExceptionCollector();
     bool hasThrown() const;
-    void rethrow() const;
+    [[noreturn]] void rethrow() const;
```
Comment (makortel) [inline FWCore/Utilities/interface/ExceptionCollector.h:38]: The clang compilation warning collector missed
```
src/FWCore/Utilities/src/ExceptionCollector.cc:36:3: warning: function declared 'noreturn' should not return [-Winvalid-noreturn]
   36 |   }
      |   ^
1 warning generated.
```
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-730ac3/42230/buildClang.log

The warning is real, as the `rethrow()` function has a code path that returns (`nExceptions_ == 0`)
https://github.com/cms-sw/cmssw/blob/1637e96339168f5237472a371cccf275ecf45ee3/FWCore/Utilities/src/ExceptionCollector.cc#L30-L36

Comment (iarspider) [reply FWCore/Utilities/interface/ExceptionCollector.h]: @makortel thanks!

**Comment by iarspider** [comment]
please test

**Comment by iarspider** [comment]
@cmsbuild ignore tests-rejected with ib-failure
Failing unit test should be fixed with https://github.com/cms-sw/cmssw/pull/46414

**Comment by makortel** [comment]
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/46416

**Comment by makortel** [comment]
+core

**Comment by mandrenguyen** [comment]
+1

**Comment by smuzaffar** [comment]
> > The clang compilation warning collector missed
> 
> Do we know why the warning wasn't flagged by the PR test machinery?

@makortel , bot flags only those warnings which are coming from files touched by the PR. In this case warning was coming from `FWCore/Utilities/src/ExceptionCollector.cc` which was not touched by this PR. We can update bot to report warning if file which has the warning is directly or in-directly using the files touched by PR ... does this sound reasonable?

**Comment by makortel** [comment]
> > > The clang compilation warning collector missed
> > 
> > 
> > Do we know why the warning wasn't flagged by the PR test machinery?
> 
> @makortel , bot flags only those warnings which are coming from files touched by the PR. In this case warning was coming from `FWCore/Utilities/src/ExceptionCollector.cc` which was not touched by this PR. We can update bot to report warning if file which has the warning is directly or in-directly using the files touched by PR ... does this sound reasonable?

Sounds good, thanks.

**Pull request merged**